### PR TITLE
chore(hogql): Add retention queries in HogQL

### PR DESCRIPTION
**Pull request by webjunkie**
## Problem

We want to convert the retention queries to use HogQL as well.

## Changes

- make new retention query runner

## How did you test this code?

- test in frontend using feature flag
- copied tests from old retention query and adapted them, made all pass

### CODE REVIEW

**Comment by webjunkie** [comment]
@mariusandra This is a bit rough from the code maybe, but mostly working. The tests fail at some poe stuff. Let's maybe discuss that tomorrow. Otherwise: Did I miss anything big?

File: posthog/models/team/team.py [hunk 1]
```diff
@@ -360,7 +360,7 @@ def groups_seen_so_far(self, group_type_index: GroupTypeIndex) -> int:
             {"team_id": self.pk, "group_type_index": group_type_index},
         )[0][0]
 
-    @cached_property
+    @property
```
Comment (mariusandra) [inline posthog/models/team/team.py:363]: Hmm, why this change?

File: frontend/src/queries/schema.ts [hunk 2]
```diff
@@ -498,8 +498,13 @@ export interface FunnelsQuery extends InsightsQueryBase {
 
 /** `RetentionFilterType` minus everything inherited from `FilterType` */
 export type RetentionFilter = Omit<RetentionFilterType, keyof FilterType>
+
+export interface RetentionQueryResponse extends QueryResponse {
+    results: any[]
```
Comment (mariusandra) [inline frontend/src/queries/schema.ts:503]: Would be great to have this a bit more complete

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+    RetentionQueryResponse,
+    IntervalType,
+)
+from posthog.schema import RetentionQuery, RetentionType
+
+
+class RetentionQueryRunner(QueryRunner):
+    query: RetentionQuery
+    query_type = RetentionQuery
+    EVENT_TABLE_ALIAS = "e"
```
Comment (mariusandra) [inline posthog/hogql_queries/insights/retention_query_runner.py:43]: I don't think we need this and can just refer to the table as "events".

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+        query: RetentionQuery | Dict[str, Any],
+        team: Team,
+        timings: Optional[HogQLTimings] = None,
+        modifiers: Optional[HogQLQueryModifiers] = None,
+        in_export_context: Optional[bool] = None,
+        event_query_type: Optional[RetentionQueryType] = None,
+    ):
+        super().__init__(query, team=team, timings=timings, modifiers=modifiers, in_export_context=in_export_context)
+        self.event_query_type = event_query_type
+        self.old_filter = RetentionFilter(data=self.query.retentionFilter.model_dump())
```
Comment (mariusandra) [inline posthog/hogql_queries/insights/retention_query_runner.py:56]: We'll want to change the system such that we're not referencing the old filter classes/code anymore. Ideally we should/could get what we need out of `self.query.retentionFilter` directly.

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+        *,
+        source: str = "timestamp",
+        ensure_datetime: bool = False,
+    ) -> str:
+        trunc_func = get_trunc_func_ch(self.query.retentionFilter.period.name.lower())
+        trunc_func_args = [source]
+        if trunc_func == "toStartOfWeek":
+            trunc_func_args.append((WeekStartDay(self.team.week_start_day or 0)).clickhouse_mode)
+        interval_sql = f"{trunc_func}({', '.join(trunc_func_args)})"
+        return interval_sql
```
Comment (mariusandra) [inline posthog/hogql_queries/insights/retention_query_runner.py:69]: Perhaps `QueryDateRange` could be used for these date manipulations? Perhaps not... not sure.

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+            else self.query_date_range.date_to()
+        )
+        if self.query.retentionFilter.period != "Hour":
+            start_date = start_date.replace(hour=0, minute=0, second=0, microsecond=0)
+            end_date = end_date.replace(hour=0, minute=0, second=0, microsecond=0)
+        params = {
+            f"{self.event_query_type}_start_date": start_date,
+            f"{self.event_query_type}_end_date": end_date,
+        }
+        return query, params
```
Comment (mariusandra) [inline posthog/hogql_queries/insights/retention_query_runner.py:210]: I think we want to avoid passing around an extra `params`... though maybe it's fine, I have little context 🤔 :D

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+    def build_returning_event_query(self):
+        runner = RetentionQueryRunner(
+            query=self.query,
+            team=self.team,
+            timings=self.timings,
+            modifiers=self.modifiers,
+            in_export_context=self.in_export_context,
+            event_query_type=RetentionQueryType.RETURNING,
+        )
+        return runner.retention_events_query()
```
Comment (mariusandra) [inline posthog/hogql_queries/insights/retention_query_runner.py:236]: It's kind of weird to have this self-referencing class. Are there perhaps two separate things? Or could this just call `self.make_query("returning")` or so? 🤔

**Review by mariusandra — COMMENTED** [review]
Didn't get to the end yet, but sharing the comments I already have.

File: posthog/hogql_queries/insights/retention_query_runner.py (hunk 5 above)
Comment (Gilbert09) [reply posthog/hogql_queries/insights/retention_query_runner.py:69]: +1 on using `QueryDateRange` here. It'll give you the interval trunc function etc

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+            _fields += [
+                f"""
+                [
+                    dateDiff(
+                        {{period}},
+                        {self.get_start_of_interval_sql(source='{start_date}')},
+                        {self.get_start_of_interval_sql(source=source_timestamp)}
+                    )
+                ] as breakdown_values
+                """
```
Comment (Gilbert09) [inline posthog/hogql_queries/insights/retention_query_runner.py:92]: I'd advise against using string interpolation for any of the SQL stuff. It makes the SQL much harder to read and becomes unmaintainable when you're just passing ambiguous `strings` around. I'd try to construct these into actual `AST` nodes, whether thats an `ast.call(..)` for the `dateDiff` and all its parameters. Here's a good doc for getting started with the ast system - https://posthog.com/handbook/engineering/databases/hogql-python

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+        ensure_datetime: bool = False,
+    ) -> str:
+        trunc_func = get_trunc_func_ch(self.query.retentionFilter.period.name.lower())
+        trunc_func_args = [source]
+        if trunc_func == "toStartOfWeek":
+            trunc_func_args.append((WeekStartDay(self.team.week_start_day or 0)).clickhouse_mode)
+        interval_sql = f"{trunc_func}({', '.join(trunc_func_args)})"
+        return interval_sql
+
+    def retention_events_query(self):
```
Comment (Gilbert09) [inline posthog/hogql_queries/insights/retention_query_runner.py:71]: An explicit return type on all funcs are very helpful 🙏

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+                ] as breakdown_values
+                """
+            ]
+
+        params.update(
+            {
+                "start_date": self.query_date_range.date_from(),
+                "period": self.query_date_range.interval_name,
+            }
+        )
```
Comment (Gilbert09) [inline posthog/hogql_queries/insights/retention_query_runner.py:100]: with things like `parse_expr`, you can still pass placeholders in if need be, which can avoid having these param objects floating about

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+                **hogql_params,
+                "filter_where": ast.And(exprs=filter_expressions)
+                if len(filter_expressions) > 1
+                else filter_expressions[0],
+            },
+            timings=self.timings,
+        )
+        result.where = ast.And(exprs=filter_expressions)
+        result.group_by = group_by_fields
+        result.having = having_expr
```
Comment (Gilbert09) [inline posthog/hogql_queries/insights/retention_query_runner.py:145]: I'd opt for using `ast.Field` nodes instead of the `_fields` join here. I'd personally also add the where/group_by/having into the SQL statement directly and have these in the placeholders, e.g:
```sql
SELECT {fields}
FROM events
WHERE {where_clause}
GROUP BY {group_by_fields}
HAVING {having_expr}
```

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+            return action_to_expr(
+                action,
+            )
+        elif entity.type == TREND_FILTER_TYPE_EVENTS:
+            if entity.id is None:
+                return ast.Constant(value=True)
+            else:
+                return ast.CompareOperation(
+                    op=ast.CompareOperationOp.Eq,
+                    left=ast.Field(chain=[self.EVENT_TABLE_ALIAS, "event"]),
```
Comment (Gilbert09) [inline posthog/hogql_queries/insights/retention_query_runner.py:180]: Yeah, the `EVENT_TABLE_ALIAS` isn't really needed here. You can let HogQL do this for you, or, drop the alias completely if possible.

File: posthog/hogql_queries/insights/retention_query_runner.py (hunk 6 above)
Comment (Gilbert09) [inline posthog/hogql_queries/insights/retention_query_runner.py:210]: Can we use `QueryDateRange` here? It performs a lot of logic around setting the timestamps on dates dependent on the interval set etc, which probably covers a few more use cases than here

File: posthog/hogql_queries/insights/retention_query_runner.py (hunk 7 above)
Comment (webjunkie) [reply posthog/hogql_queries/insights/retention_query_runner.py:236]: The problem was with the `event_query_type` being set (for the whole instance) and this then being referenced like 20 times. And I didn't want to pass it through everywhere. Hence I made instances.
I can check once more, maybe it's now at this point gotten simpler to achieve.

File: posthog/hogql_queries/insights/retention_query_runner.py (hunk 8 above)
Comment (webjunkie) [reply posthog/hogql_queries/insights/retention_query_runner.py:92]: I see. I kinda started on that but it's really pain with nested calls and functions and references to fields or aliases and whatnot.

Comment (Gilbert09) [reply posthog/hogql_queries/insights/retention_query_runner.py:92]: The alternative is to use `parse_expr` which also works

File: posthog/models/team/team.py (hunk 1 above)
Comment (webjunkie) [reply posthog/models/team/team.py:363]: It tripped up the tests. The test uses a team, asserts something, then sets the team timezone, and tries to assert once more. But due to the timezone property that the test uses being cached here on the instance, it didn't reflect the change. Not sure why it worked in the old tests.

But, I wouldn't recommend to put `cached_property` on "simple" properties like these anyways, since they are directly computed from available data and do not need to be queried for or such. It's often times more a source of confusion and stale data than helping performance. Hence I removed it. What do you think?

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+                op=ast.CompareOperationOp.Eq,
+                left=ast.Field(chain=["events", "event"]),
+                right=ast.Constant(value=entity["id"]),
+            )
+
+        return ast.CompareOperation(
+            op=ast.CompareOperationOp.Eq,
+            left=ast.Field(chain=["events", "event"]),
+            right=ast.Constant(value=PAGEVIEW_EVENT),
+        )
```
Comment (mariusandra) [inline posthog/hogql_queries/insights/retention_query_runner.py:161]: I think this could be some utility function in hogql/property.py, as I think both trends and lifecycle have somewhat identical code somewhere.

File: posthog/queries/test/__snapshots__/test_lifecycle.ambr [hunk 14]
```diff
@@ -148,8 +148,8 @@
   '
   WITH 'day' AS selected_period,
        periods AS
-    (SELECT dateSub(day, number, dateTrunc(selected_period, toDateTime('2020-01-19 23:59:59', 'US/Pacific'))) AS start_of_period
-     FROM numbers(dateDiff('day', dateTrunc('day', toDateTime('2020-01-12 00:00:00', 'US/Pacific')), dateTrunc('day', toDateTime('2020-01-19 23:59:59', 'US/Pacific') + INTERVAL 1 day))))
+    (SELECT dateSub(day, number, dateTrunc(selected_period, toDateTime('2020-01-18 23:59:59', 'US/Pacific'))) AS start_of_period
+     FROM numbers(dateDiff('day', dateTrunc('day', toDateTime('2020-01-11 00:00:00', 'US/Pacific')), dateTrunc('day', toDateTime('2020-01-18 23:59:59', 'US/Pacific') + INTERVAL 1 day))))
```
Comment (mariusandra) [inline posthog/queries/test/__snapshots__/test_lifecycle.ambr:152]: This seems like a bug? The date ranges in the lifecycle tests do seem to be 12-19, not 11-18... 🤔

File: posthog/queries/test/__snapshots__/test_retention.ambr [hunk 15]
```diff
@@ -416,7 +416,7 @@
                         [
                           dateDiff(
                               'Day',
-                              toStartOfDay(toDateTime('2020-06-10 00:00:00', 'US/Pacific')),
+                              toStartOfDay(toDateTime('2020-06-10 07:00:00', 'US/Pacific')),
```
Comment (mariusandra) [inline posthog/queries/test/__snapshots__/test_retention.ambr:419]: Without digging in deeper, this seems like a bug?

**Review by mariusandra — COMMENTED** [review]
Looking good. I didn't run it yet, just read through the code. Left a few comments inline.

File: posthog/hogql_queries/insights/retention_query_runner.py (hunk 13 above)
Comment (webjunkie) [reply posthog/hogql_queries/insights/retention_query_runner.py:161]: I didn't find code that does it like this, or are you suggesting to move it there?

Comment (mariusandra) [reply posthog/hogql_queries/insights/retention_query_runner.py:161]: This seems like something that'll be common to all insights, as most offer you an "entity" (action vs event) to choose from. It's not yet abstracted.

File: posthog/queries/test/__snapshots__/test_lifecycle.ambr (hunk 14 above)
Comment (webjunkie) [reply posthog/queries/test/__snapshots__/test_lifecycle.ambr:152]: 12-19 00:00Z is 12-18 in US/Pacific.

I confirmed that the this now just reflects how the date calculations work: Send a query from frontend that has e.g. `date_to: 2023-12-01T06:00:00.000Z` while having US/Pacfic time on the team, and the SQL query in the console will show the day before.

File: frontend/src/queries/schema.ts [hunk 16]
```diff
@@ -512,10 +512,27 @@ export interface FunnelsQuery extends InsightsQueryBase {
...
+    /** @asType integer */
+    count: number
+}
+
+export type RetentionResult = {
+    values: RetentionValue[]
+    label: string
+    /** @format date-time */
+    date: string
+}
```
[the remark this answers is not in the document]
Comment (webjunkie) [reply frontend/src/queries/schema.ts:526]: What do we prefer?

[the remark this answers is not in the document]
Comment (mariusandra) [reply frontend/src/queries/schema.ts:526]: Well, everything that's not a oneliner is an interface, so best to stick to the pattern...?

File: posthog/hogql/property.py [hunk 17]
```diff
@@ -381,6 +387,27 @@ def action_to_expr(action: Action) -> ast.Expr:
...
+            op=ast.CompareOperationOp.Eq,
+            left=ast.Field(chain=["events", "event"]),
+            right=ast.Constant(value=entity["id"]),
+        )
+
+    return ast.CompareOperation(
+        op=ast.CompareOperationOp.Eq,
+        left=ast.Field(chain=["events", "event"]),
+        right=ast.Constant(value=default_event),
+    )
```
Comment (mariusandra) [inline posthog/hogql/property.py:408]: Would be great to have a test or two for this.

File: posthog/queries/test/__snapshots__/test_retention.ambr (hunk 15 above)
Comment (mariusandra) [reply posthog/queries/test/__snapshots__/test_retention.ambr:419]: Since the regression to these lines happens in this PR, I'm also fine with the fix in here if it's a oneliner. I saw another "00:00" -> "16:00" in `test_lifecycle_query_runner.ambr`. If you think it's easier, then I have nothing against an additional PR either...

File: posthog/queries/test/test_lifecycle.py [hunk 18]
```diff
@@ -25,7 +25,17 @@ def create_action(**kwargs):
...
+class TestLifecycleBase(ClickhouseTestMixin, APIBaseTest):
+    def assertLifecycleResults(self, results, expected):
+        sorted_results = [
+            {"status": r["status"], "data": r["data"]} for r in sorted(results, key=lambda r: r["status"])
+        ]
+        sorted_expected = list(sorted(expected, key=lambda r: r["status"]))
+
+        self.assertListEqual(sorted_results, sorted_expected)
+
+
```
Comment (mariusandra) [inline posthog/queries/test/test_lifecycle.py:37]: I'm not sure why the lifecycle query had to change in this retention query PR? Something to do with the query date range?

File: posthog/queries/test/test_lifecycle.py [hunk 19]
```diff
@@ -881,26 +890,18 @@ def test_timezones(self):
...
                 {
                     "status": "dormant",
-                    "data": [-1.0, -2.0, -1.0, 0.0, -2.0, 0.0, -1.0, 0.0],
+                    "data": [0.0, -1.0, -2.0, -1.0, 0.0, -2.0, 0.0, -1.0],
                 },
-                {"status": "new", "data": [1, 0, 0, 1, 0, 0, 0, 0]},
-                {"status": "resurrecting", "data": [1, 1, 0, 1, 0, 1, 0, 1]},
+                {"status": "new", "data": [1, 1, 0, 0, 1, 0, 0, 0]},
+                {"status": "resurrecting", "data": [0, 1, 1, 0, 1, 0, 1, 0]},
                 {"status": "returning", "data": [0, 0, 0, 0, 0, 0, 0, 0]},
```
Comment (mariusandra) [inline posthog/queries/test/test_lifecycle.py:905]: How come the query results changed in an old non-hogql insight?

Ideally we want to completely `rm -rf` this entire `posthog/queries` folder, or at least the insights contained within. Modifying those existing tests seems like a bit of wasted effort... or I'm likely just not seeing it 🤔

File: posthog/hogql_queries/utils/query_date_range.py [hunk 20]
```diff
@@ -240,3 +241,62 @@ def to_properties(self, field: Optional[List[str]] = None) -> List[ast.Expr]:
...
+        return date_to.replace(hour=0, minute=0, second=0, microsecond=0)
+
+    def get_start_of_interval_hogql(self, *, source: ast.Expr = None) -> ast.Expr:
+        trunc_func = get_trunc_func_ch(self._interval.name.lower())
+        trunc_func_args = [source] if source else [ast.Constant(value=self.date_from())]
+        if trunc_func == "toStartOfWeek":
+            trunc_func_args.append(
+                ast.Constant(value=int((WeekStartDay(self._team.week_start_day or 0)).clickhouse_mode))
+            )
+        return ast.Call(name=trunc_func, args=trunc_func_args)
```
Comment (mariusandra) [inline posthog/hogql_queries/utils/query_date_range.py:302]: All of these should also get tests.

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 21]
```diff
@@ -0,0 +1,316 @@
...
+        in_export_context: Optional[bool] = None,
+    ):
+        super().__init__(query, team=team, timings=timings, modifiers=modifiers, in_export_context=in_export_context)
+        if not self.query.retentionFilter:
+            self.query.retentionFilter = RetentionFilter()  # TODO: Clarify default filter
+
+        self.target_entity = self.query.retentionFilter.target_entity or {
+            "id": "$pageview",
+            "type": TREND_FILTER_TYPE_EVENTS,
+        }
```
Comment (mariusandra) [inline posthog/hogql_queries/insights/retention_query_runner.py:54]: I know you didn't modify `renteintionFilter`, but it's a bit weird to see a dict, when everything else operates on dataclasses. Now that you know what shape the data takes here, can you add the correct types into the old retention filter type:
https://github.com/PostHog/posthog/blob/55c7f08d61f6378cddb1162353adb1677c3d4e9d/frontend/src/types.ts#L1882-L1883

File: posthog/hogql_queries/insights/retention_query_runner.py [hunk 22]
```diff
@@ -0,0 +1,316 @@
...
+
+        group_by_fields = None
+        having_expr = None
+        if event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+            group_by_fields = [ast.Field(chain=["target"])]
+            having_expr = date_filter_expr
+        elif event_query_type == RetentionQueryType.RETURNING:
+            group_by_fields = [ast.Field(chain=["target"]), ast.Field(chain=["event_date"])]
+
+        result = parse_select("SELECT * FROM events", timings=self.timings)
```
Comment (mariusandra) [inline posthog/hogql_queries/insights/retention_query_runner.py:125]: Nit: you can probably get by here by just writing the AST manually. Feel free to ignore, as this is fine too.

**Review by mariusandra — CHANGES_REQUESTED** [review]
Great progress. I left some inline comments, though the main things still pending:

1. Date range data integrity. Things seem to have changed where they shouldn't have.
2. Some of the utilities/classes could use tests.
3. I'm also not sure why did you have to change the lifecycle query, and why did its test/results change? There's probably a good reason 😅

File: posthog/hogql_queries/insights/retention_query_runner.py (hunk 21 above)
Comment (webjunkie) [reply posthog/hogql_queries/insights/retention_query_runner.py:54]: You mean to use also pydantic dataclasses here? I also thought that would be useful and I think I checked and it trickles down quite a lot... would be additional work, but can be done 👍🏻

File: posthog/hogql_queries/insights/retention_query_runner.py (hunk 22 above)
Comment (webjunkie) [reply posthog/hogql_queries/insights/retention_query_runner.py:125]: Yeah I had attempted something like that and it didn't work here. Will find out what it was.

File: posthog/queries/test/__snapshots__/test_retention.ambr (hunk 15 above)
Comment (webjunkie) [reply posthog/queries/test/__snapshots__/test_retention.ambr:419]: Put into #19088

File: frontend/src/queries/nodes/InsightQuery/utils/filtersToQueryNode.test.ts [hunk 23]
```diff
@@ -167,7 +167,9 @@ describe('filtersToQueryNode', () => {
             const query: InsightQueryNode = {
                 kind: NodeKind.RetentionQuery,
                 filterTestAccounts: true,
+                retentionFilter: {},
             }
+            result.retentionFilter = {} // Put this here to make TS happy
```
Comment (mariusandra) [inline frontend/src/queries/nodes/InsightQuery/utils/filtersToQueryNode.test.ts:172]: It seems to make TS unhappy...
I don't understand why we need it. Shouldn't `filtersToQueryNode` take care of stuff like that?

**Review by mariusandra — APPROVED** [review]
I think this looks good! Great work! 👍 

Still some tests / types to fix though, and then the persons modal stuff... but otherwise, let's get it in!

File: frontend/src/queries/nodes/InsightQuery/utils/filtersToQueryNode.test.ts (hunk 23 above)
Comment (webjunkie) [reply frontend/src/queries/nodes/InsightQuery/utils/filtersToQueryNode.test.ts]: I have the query typed to require `retentionFilter`, hence added it, but since it's empty the result var didn't contain it and the equal test was failing. Anyway, I just used a cast to bypass issues now.
